SI-6646 Fix regression in for desugaring.

retronym opened the pull request:
The early check in the parser of pattern irrefutability,
added in c82ecab, failed to consider InitCaps and
`backquoted` identifiers.

Review by @paulp

@adriaanm should take a look to judge the severity.

Given that it's a silent killer, I've formed an opinion that is
expressed with the target branch of this PR.

retronym committed:
SI-6646 Fix regression in for desugaring.

The early check in the parser of pattern irrefutability,
added in c82ecab, failed to consider InitCaps and
`backquoted` identifiers.

jsuereth commented:
I believe, given the spec and that this is regression, it's a blocker.

paulp commented:
Looks like you're cleaning up after me all over the place.  LGTM.

retronym commented:
:cat: https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/663/console

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1615/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1615/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/905/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/905/

retronym committed:
SI-6646 `ident` or Ident is always new binding.

The previous commit regressed in these cases:

    // no withFilter
    for (X <- List("A single ident is always a pattern")) println(X)
    for (`x` <- List("A single ident is always a pattern")) println(`x`)

At the top level of the LHS of a <-, such identifiers represent
new bindings, not stable identifier patterns.

retronym commented:
Review once more by @paulp, please.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/923/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/923/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1629/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1629/

paulp commented on src/reflect/scala/reflect/internal/TreeInfo.scala:
```diff
@@ -245,6 +245,57 @@ abstract class TreeInfo {
+   * in the position `for { <tree> <- expr }` based only
+   * on information at the `parser` phase?
+   *
+   * This check is not based on
```
this check is not based on... WHAT, THEY'RE ENDING THE SEASON THAT WAY?

retronym replied:
The producers have decided this was too controversial and have rounded out the story.

paulp commented on src/reflect/scala/reflect/internal/TreeInfo.scala:
```diff
@@ -245,6 +245,57 @@ abstract class TreeInfo {
+   * introduce new bindings:
+   *
+   * {{{
+   * for { X <- xs } yield X
```
https://groups.google.com/forum/?fromgroups=#!topic/scala-internals/qwa_XOw_7Ks

retronym replied:
I've added the link to the comment. Makes (enough) sense:  `__ <-` is like `__ =`, `case __  =>` is the odd one out.

retronym committed:
Update comment.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/931/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/931/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1636/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1636/

paulp commented:
Looks good to me after our exciting little side trip to scala-internals.

adriaanm commented:
Excellent. I agree this should be fixed now. Great docs.
